Bio/Align/mauve.py

```diff
@@ -0,0 +1,233 @@
+# Copyright 2015-2015 by Eric Rasche.  All rights reserved.
+#
+# This file is part of the Biopython distribution and governed by your
+# choice of the "Biopython License Agreement" or the "BSD 3-Clause License".
+# Please see the LICENSE file that should have been included as part of this
+# package.
+"""Bio.Align support for "xmfa" output from Mauve/ProgressiveMauve.
+
+You are expected to use this module via the Bio.Align functions.
+"""
+
+from Bio.Align import interfaces, Alignment
+from Bio.Seq import Seq, reverse_complement
+from Bio.SeqRecord import SeqRecord
+
+
+class AlignmentWriter(interfaces.AlignmentWriter):
+    """Mauve/XMFA alignment writer."""
+
+    def write_header(self, alignments):
+        """Write the file header to the output file."""
+        stream = self.stream
+        metadata = alignments.metadata
+        format_version = metadata.get("FormatVersion", "Mauve1")
+        line = f"#FormatVersion {format_version}\n"
+        stream.write(line)
+        alignment = alignments[0]
+        # first alignment always seems to contain all sequences
+        identifiers = [sequence.id for sequence in alignment.sequences]
+        filename = metadata.get("File")
+        if filename is None:
+            # sequences came from separate files
+            for index, filename in enumerate(identifiers):
+                number = index + 1
+                line = f"#Sequence{number}File\t{filename}\n"
+                stream.write(line)
+                line = f"#Sequence{number}Format\tFastA\n"
+                stream.write(line)
+            self._filenames = identifiers
+        else:
+            # sequences came from one combined file
+            for number, identifier in enumerate(identifiers):
+                assert number == int(identifier)
+                number += 1
+                line = f"#Sequence{number}File\t{filename}\n"
+                stream.write(line)
+                line = f"#Sequence{number}Entry\t{number}\n"
+                stream.write(line)
+                line = f"#Sequence{number}Format\tFastA\n"
+                stream.write(line)
+        backbone_file = metadata.get("BackboneFile", None)
+        if backbone_file is not None:
+            line = f"#BackboneFile\t{backbone_file}\n"
+            stream.write(line)
+
+    def write_file(self, alignments):
+        """Write a file with the alignments, and return the number of alignments.
+
+        alignments - A Bio.Align.mauve.AlignmentIterator object.
+        """
+
+        class ListWithAttributes(list):
+            pass
+
+        try:
+            metadata = alignments.metadata
+        except AttributeError:
+            metadata = {}
```

Comment:
No test coverage, but can that fail with the object model?

Reply:
It cannot fail if the alignments came from Mauve originally. It could fail if the alignments came from some other source, but I doubt this module will be used to convert other alignment formats to Mauve.

Reply:
So these alignment objects will not always have ``.metadata``?

I was thinking it would be like the SeqRecord and old MSA object which always has a ``.annotations`` dictionary even if it was empty.

Reply:
The `metadata` attribute is on the `alignments` variable, which is an `AlignmentIterator` object.

Reply:
And that class is defined in ``Bio/Align/interfaces.py`` and does not have any ``.metadata`` attribute, so the Mauve parser adds it at run time? That feels like magic to me (but explains the try/except AttributeError).

Reply:
There is nothing wrong with adding an attribute to a class instance at run time. See this example in the Python documentation:

https://docs.python.org/3/tutorial/classes.html#instance-objects

Reply:
I know Python lets you add attributes at run time, but for a uniformed alignment parsing framework (like we had with AlignIO) we need to store metadata consistently, especially if any of it is directly transferable between file formats.

Different parsers adding different attributes at run time is harder to keep track of and document.

Admittedly with SeqIO and AlignIO we don't document many of the keys used, but the ``.annotations`` dictionary (and similar dictionaries for per-column metadata etc) are documented and used consistently. It was clear where to start looking (rather than here, where you'd have to try run-time object inspection).

I would be much happier here if the ``AlignmentIterator`` had a pre-declared and documented attribute. Presumably this is for file-level annotation?

Reply:
To be precise, I am not adding attributes at run time. The design is as follows:

- `Bio.Align.interfaces` defines the `AlignmentIterator` base class;
- `Bio.Align.mauve` defines the `mauve.AlignmentIterator` class that derives from the base class and adds `.metadata` for file-level annotations as an attribute in its `__init__` method.

We are using the same approach in `Bio.SeqIO`:

- `Bio.SeqIO.interfaces` defines the `SequenceIterator` base class
- `Bio.SeqIO.SeqXmlIO` defines the `SeqXmlIterator` class that derives from base class, and adds the file-level attributes `seqXMLversion`, `source`, `sourceVersion`, `ncbiTaxID`, `speciesName`.

Reply:
Those details are helpful, thank you. The ``SeqXmlIterator`` example is recent too.

I would still prefer having a common ``.annotations`` (or ``.metadata``) dictionary, but this just moves the file-format-specific details from being attributes to different dictionary keys.

So in this specific case, depending which subclass of ``AlignmentIterator`` is used, there may or may not be a ``.metadata`` present (much like with the dictionary approach, a key may or may not be present).

Reply:
> So in this specific case, depending which subclass of `AlignmentIterator` is used, there may or may not be a .metadata present (much like with the dictionary approach, a key may or may not be present).

Yes. However, `alignments` does not necessarily need to be an `AlignmentIterator` instance. A user may for example call `write_file` on a list of `Alignment` objects.